design/automatic-network-configuration.md

```diff
@@ -0,0 +1,801 @@
+# Automatic network configuration
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal is to expand the scope of Metal³ to include an API to
+manage physical network devices.
+
+## Motivation
+
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI),
+which is an approach to use Kubernetes to manage underlying infrastructure.
+Managing the configuration of some physical network devices is closely
+related to managing physical hosts. As bare metal hosts are provisioned
+or later repurposed, there may be corresponding physical network changes
+that must be made, such as reconfiguring a ToR switch port. If the provisioning
+of the physical host is managed through a Kubernetes API, it would be
+convenient to be able to reconfigure related network devices using a
+similar API.
+
+## Goals
+
+- Define a Kubernetes API for configuring network device.
+- Automatically configure the network infrastructure for a host when
+adding it to a cluster.
+- Do network configuration when deprovisioning the hosts.
+- Design a network abstraction that can represent any of the target networking
+configuration, independently of the controller used.
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation in the future.
+- Implement a solution for lldp in the future.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost. First we
+added the NetworkConfiguration operator to implement the network configuration
```

Comment:
Where will the new operator live? I'd prefer it to be a separate repository to avoid overcomplicating BMO.

Reply:
We also intend to create a separate repository for the new operator and make sure to make the least impact to CAPM3 and BMO.